tool/src/org/antlr/v4/codegen/target/JavaTarget.java

```diff
@@ -20,6 +19,11 @@
 import java.util.Set;
 
 public class JavaTarget extends Target {
+	public final static String key = "Java";
```

Comment:
Seems like we should automate this to use the required convention we have that XTarget uses X as key.  I should have done this earlier rather than hardcoding.

Reply:
I like automatization but I don't how to implement it in a better way with Java. The key value is being checked below in the static constructor (it should be X from XTarget). These constants are used in the following switch: https://github.com/antlr/antlr4/pull/3349/files#diff-0bbbf52a2f6fceb14a08812a3c96f56f6069ba4a8acab5032ac75ca5c1b0c181R24-R39

Reply:
let me experiment. Might be another reason to avoid the switch.

Reply:
Without the switch, some logic is duplicated across several targets, that I'd like to avoid:

```java
class JavaTarget {
    ...
    protected override void appendUnicodeEscapedCodePoint(codePoint) {
        if (Character.isSupplementaryCodePoint(codePoint)) {
			// char is not an 'integral' type, so we have to explicitly convert
			// to int before passing to the %X formatter or else it throws.
			sb.append(String.format("\\u%04X", (int)Character.highSurrogate(codePoint)));
			sb.append(String.format("\\u%04X", (int)Character.lowSurrogate(codePoint)));
		}
		else {
			sb.append(String.format("\\u%04X", codePoint));
		}
    }
}

class JavaScriptTarget {
    ...
    protected override void appendUnicodeEscapedCodePoint(codePoint) {
        if (Character.isSupplementaryCodePoint(codePoint)) {
			// char is not an 'integral' type, so we have to explicitly convert
			// to int before passing to the %X formatter or else it throws.
			sb.append(String.format("\\u%04X", (int)Character.highSurrogate(codePoint)));
			sb.append(String.format("\\u%04X", (int)Character.lowSurrogate(codePoint)));
		}
		else {
			sb.append(String.format("\\u%04X", codePoint));
		}
    }
}
```

The common code can be extracted into a separate method:

```java
class JavaTarget {
    ...
    protected override void appendUnicodeEscapedCodePoint(codePoint) {
        appendUnicodeEscapedCodePointForJavaAndJavaScript(codePoint);
    }
}

class JavaScriptTarget {
    ...
    protected override void appendUnicodeEscapedCodePoint(codePoint) {
        appendUnicodeEscapedCodePointForJavaAndJavaScript(codePoint);
    }
}
```

but it's not clear how to name it. `appendUnicodeEscapedCodePointForJavaAndJavaScript` does not look clear.

Reply:
Certainly this removes need to specify "Java" etc... in subclasses:

```java
protected Target(CodeGenerator gen) {
  ...
  String clazz = this.getClass().getSimpleName();
  this.language = clazz.substring(0,clazz.indexOf("Target"));
```

but can't do string switch. :(

Reply:
`public final static String language = "Java";` is probably ok in each target to get constant.  What does it look like to remove refs to virtual method? We would replace it with direct calls to appendEscapedCodePoint().

Reply:
Yes, I've also tried similar tricks, but switch accepts only with constants (final static) :(

Reply:
> What does it look like to remove refs to virtual method? We would replace it with direct calls to appendEscapedCodePoint().

I am afraid it could break compatibility with other unofficial targets because it will not be possible to override escaping behavior. It's not possible to call an instance method from a static method.

Reply:
Ok, maybe we just add a comment to the virtual method to indicate we are leaving it in for flexibility and backward compatibility with external targets but that target implementations are free to call the static method as a support function if they like.

Reply:
per my other comment made just now, seems like we should use `case "Java":` here :)

Reply:
I forgot that these constants are also already used in tests besides `switch` statement: https://github.com/antlr/antlr4/pull/3349/files#diff-453e4cf66eb5c1ed759da1d5a64049c1307117fc10c10e0b2fc4a2dcec17b9e1R27 Should I also replace `JavaTarget.key` with `Java` there? I still think it's error-prone to use literals instead of string constants when it's possible, but I accept your point here.

Reply:
Yep, let's converted to a string and simplify this PR.  You can tease me later when we decide we need a constant here. haha.